### ISSUE TITLE
extending dictionaries

### ISSUE THREAD

**Issue by knezi** [AnySoftKeyboard/AnySoftKeyboard#2005]
Copying issue https://github.com/AnySoftKeyboard/LanguagePack/issues/352 because of the ongoing migration.

Czech dictionary is very unusable due to the insufficient size of the dictionary. Often, words are present, but not in all forms (an example https://en.wikipedia.org/wiki/Czech_declension), which makes it very hard to use word completion and autocorrection.

The ASK Czech dictionary contains approx 200K words, whereas aspell dictionary generates almost 5M. I realise that aspell may be overgenerating, that is produce words that actually do not exist (even though I haven't found any by briefly skimming through the words).

All in all, it seems as we may be able to extend the dictionaries for many languages. This could improve the usability a lot, especially for flective languages.

The questions are:

Is 5 million entries too many? (aspell stores only the stem and then generates all forms)

Aspell is missing the frequencies of words, is that a problem?

Are there non-existent words in aspell? If so, is that such a big deal?

I can provide the data files and scripts for generating them if needbe.

**Comment by knezi** [comment]
Responding to https://github.com/AnySoftKeyboard/LanguagePack/issues/352#issuecomment-581039459:
> Word frequency is very important for word suggestion. Do you think you can come up with a way to generate or guess the frequency of a word?

We could use publicly available corpora to at least estimate the frequencies. It should be sufficient and easily done IMO.
Also, we could drop the lowest frequency words to fit in the dictionary limit.

However, I think the limit 200K still needs to be raised as it is insufficient for flective languages like Czech.

**Comment by ftyers** [comment]
The dictionaries could used compiled finite-state machines that are weighted per wordform using existing corpora. A finite-state machine is basically a DAG and fairly easy to implement. For larger languages, (character) neural language models could also be used.

**Comment by knezi** [comment]
@ftyers agreed, this is a must for inflectional languages. Unfortunately, I won't have time to look into this for at least a couple of months.

**Comment by ftyers** [comment]
My suggestion would be to use some existing FST library, for example HFST, where they have some Java code for doing "lookup" over optimised binary FSTs. https://github.com/hfst/hfst-optimized-lookup/tree/master/hfst-optimized-lookup-java 

Could you give me some pointers, perhaps I could look at it?

**Comment by knezi** [comment]
I'm not familiar myself with the source code of AnySoftKeyboard. That's why it'll take me a while to get to it.
Perhaps @menny can point you somewhere?
